druid/examples/image.rs

```diff
@@ -12,44 +12,230 @@
 };
+use druid::{AppLauncher, Color, Data, ImageBuf, Lens, LensExt, WindowDesc};
 
-pub fn main() {
-    let main_window = WindowDesc::new(ui_builder);
-    let data = 0_u32;
-    AppLauncher::with_window(main_window)
-        .use_simple_logger()
-        .launch(data)
-        .expect("launch failed");
+static FILL_STRAT_OPTIONS: &[(&str, FillStrat)] = &[
+    ("Contain", FillStrat::Contain),
+    ("Cover", FillStrat::Cover),
+    ("Fill", FillStrat::Fill),
+    ("FitHeight", FillStrat::FitHeight),
+    ("FitWidth", FillStrat::FitWidth),
+    ("None", FillStrat::None),
+    ("ScaleDown", FillStrat::ScaleDown),
+];
+
+static INTERPOLATION_MODE_OPTIONS: &[(&str, InterpolationModeData)] = &[
+    (
+        "Bilinear",
+        InterpolationModeData::new(InterpolationMode::Bilinear),
+    ),
+    (
+        "NearestNeighbor",
+        InterpolationModeData::new(InterpolationMode::NearestNeighbor),
+    ),
+];
+#[derive(Clone, Data, Lens)]
+struct AppState {
+    fill_strat: FillStrat,
+    interpolate: bool,
+    interpolation_mode: InterpolationModeData,
+    fix_width: bool,
+    width: f64,
+    fix_height: bool,
+    height: f64,
 }
 
-fn ui_builder() -> impl Widget<u32> {
+#[derive(Clone, PartialEq)]
+struct InterpolationModeData(InterpolationMode);
+
+impl InterpolationModeData {
+    const fn new(mode: InterpolationMode) -> InterpolationModeData {
+        InterpolationModeData(mode)
+    }
+}
+impl druid::Data for InterpolationModeData {
+    fn same(&self, other: &InterpolationModeData) -> bool {
+        self.0 == other.0
+    }
+}
+
+/// builds a child Flex widget from some paramaters.
+struct Rebuilder {
+    inner: Box<dyn Widget<AppState>>,
+}
+
+impl Rebuilder {
+    fn new() -> Rebuilder {
+        Rebuilder {
+            inner: SizedBox::empty().boxed(),
+        }
+    }
+
+    fn rebuild_inner(&mut self, data: &AppState) {
+        self.inner = build_widget(&data);
+    }
+}
+
+impl Widget<AppState> for Rebuilder {
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut AppState, env: &Env) {
+        self.inner.event(ctx, event, data, env)
+    }
+
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &AppState, env: &Env) {
+        if let LifeCycle::WidgetAdded = event {
+            self.rebuild_inner(data);
+        }
+        self.inner.lifecycle(ctx, event, data, env)
+    }
+
+    fn update(&mut self, ctx: &mut UpdateCtx, old_data: &AppState, data: &AppState, _env: &Env) {
+        if !old_data.same(&data) {
+            self.rebuild_inner(data);
+            ctx.children_changed();
+        }
+    }
+
+    fn layout(
+        &mut self,
+        ctx: &mut LayoutCtx,
+        bc: &BoxConstraints,
+        data: &AppState,
+        env: &Env,
+    ) -> Size {
+        self.inner.layout(ctx, bc, data, env)
+    }
+
+    fn paint(&mut self, ctx: &mut PaintCtx, data: &AppState, env: &Env) {
+        self.inner.paint(ctx, data, env)
+    }
```

Comment:
One gotcha here: if you write a widget that wraps another widget and does *not* use a `WidgetPod`, you should implement the following method; it will only matter in a small number of cases (if your inner widget has an explicit `WidgetId`) but it is strictly correct:

```suggestion
    }
    
    fn id(&self) -> Option<WidgetId> {
        self.inner.id()
    }
```

Reply:
(maybe do this yourself instead of just merging my change unseen, because I haven't tested it and I might've screwed something up)